src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+            if (node.AwaitOpt is { } awaitableInfo)
+            {
+                GetAwaitableInstancePlaceholders(placeholders, awaitableInfo, _localScopeDepth);
+            }
+
+            using var region = new PlaceholderRegion(this, placeholders);
+            placeholders.Free();
+
+            return base.VisitUsingLocalDeclarations(node);
+        }
+
+        public override BoundNode? VisitFixedStatement(BoundFixedStatement node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitFixedStatement(node);
+        }
+
+        public override BoundNode? VisitDoStatement(BoundDoStatement node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitDoStatement(node);
+        }
+
+        public override BoundNode? VisitWhileStatement(BoundWhileStatement node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitWhileStatement(node);
+        }
+
+        public override BoundNode? VisitSwitchStatement(BoundSwitchStatement node)
+        {
+            using var _ = new LocalScope(this, node.InnerLocals);
+            using var patternInput = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            base.VisitSwitchStatement(node);
+            return null;
+        }
+
+        public override BoundNode? VisitConvertedSwitchExpression(BoundConvertedSwitchExpression node)
+        {
+            using var patternInput = new PatternInput(this, GetValEscape(node.Expression, _localScopeDepth));
+            base.VisitConvertedSwitchExpression(node);
+            return null;
+        }
+
+        public override BoundNode? VisitSwitchSection(BoundSwitchSection node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchSection(node);
+        }
+
+        public override BoundNode? VisitSwitchExpressionArm(BoundSwitchExpressionArm node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitSwitchExpressionArm(node);
+        }
+
+        public override BoundNode? VisitCatchBlock(BoundCatchBlock node)
+        {
+            using var _ = new LocalScope(this, node.Locals);
+            return base.VisitCatchBlock(node);
+        }
+
+        public override BoundNode? VisitLocal(BoundLocal node)
+        {
+            Debug.Assert(_localEscapeScopes?.ContainsKey(node.LocalSymbol) == true ||
+                (node.LocalSymbol.ContainingSymbol is SynthesizedSimpleProgramEntryPointSymbol entryPoint && _symbol != entryPoint));
+
+            return base.VisitLocal(node);
+        }
+
+        private void AddLocalScopes(LocalSymbol local, uint refEscapeScope, uint valEscapeScope)
+        {
+            var scope = _useUpdatedEscapeRules ? local.Scope : DeclarationScope.Unscoped;
```

Comment:
>scope [](http://example.com/codeflow?start=16&length=5)

nit: I found the name `scope` confusing here. Same applies to existing `LocalSymbol.Scope`. I'm not sure what's a better term, maybe `scopedModifier`?